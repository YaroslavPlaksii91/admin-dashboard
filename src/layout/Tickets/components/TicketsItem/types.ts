export type TicketType = {
  id: number;
  title: string;
  customer: {
    name: string;
    date: string;
  };
  date: string;
  priority: string;
  image: string;
  updated: string;
};

export type TicketsItemProps = {
  ticket: TicketType;
};
