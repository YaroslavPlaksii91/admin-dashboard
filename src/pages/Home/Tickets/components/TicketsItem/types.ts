import { MouseEventHandler } from 'react';

export type TicketType = {
  id?: string;
  title: string;
  customerName: string;
  customerDate: string;
  date: string;
  priority: string;
  image: string;
  updated: string;
};

export type TicketsItemProps = {
  ticket: TicketType;
  handleDelete: MouseEventHandler<HTMLDivElement>;
  handleEdit: MouseEventHandler<HTMLDivElement>;
};
