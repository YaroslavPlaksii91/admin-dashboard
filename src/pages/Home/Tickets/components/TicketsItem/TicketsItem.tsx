import { FC } from 'react';
import { Grid, Button, SvgIcon, Box } from '@mui/material';

import { Label } from '@components/Label/Label';
import { LABEL_TYPES } from '@components/Label/constants';
import { getTicketUpdateTime } from '@pages/Home/Tickets/helpers/getTicketUpdateTime';
import { formatCustomerDate } from '@pages/Home/Tickets/helpers/formatCustomerDate';
import { formatDate } from '@services/date/formatDate';

import { TicketsItemProps } from './types';
import { TextCell } from '../TextCell/TextCell';

export const TicketsItem: FC<TicketsItemProps> = ({ ticket }) => {
  const updatedTime = getTicketUpdateTime(ticket.updated);
  const customerDate = formatCustomerDate(ticket.customerDate);
  const date = formatDate(ticket.date);

  return (
    <Grid
      container
      sx={{
        padding: '24px 32px',
        borderBottom: '1px solid',
        borderBottomColor: 'borderColor',
        transition: 'background-color 250ms ease',
        '&:hover': {
          backgroundColor: 'rgba(55, 81, 255, 0.04)',
          cursor: 'pointer',
        },
      }}
    >
      <Grid item xs={5} sx={{ display: 'flex', alignItems: 'center' }}>
        <Box
          sx={{
            width: '44px',
            height: '44px',
            mr: 6,
            borderRadius: '50%',
            overflow: 'hidden',
          }}
        >
          <img src={ticket.image} alt={ticket.customerName} />
        </Box>

        <TextCell text={ticket.title} subtext={updatedTime} />
      </Grid>
      <Grid item xs={3}>
        <TextCell text={ticket.customerName} subtext={`on ${customerDate}`} />
      </Grid>
      <Grid item xs={2}>
        <TextCell text={date.date} subtext={date.time} />
      </Grid>
      <Grid
        item
        xs={2}
        sx={{
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
      >
        <Label
          text={ticket.priority}
          color={ticket.priority}
          type={LABEL_TYPES.ROUND}
        />
        <Button>
          <SvgIcon titleAccess="More options">
            <use href="/src/assets/icons/sprite.svg#icon-more"></use>
          </SvgIcon>
        </Button>
      </Grid>
    </Grid>
  );
};
