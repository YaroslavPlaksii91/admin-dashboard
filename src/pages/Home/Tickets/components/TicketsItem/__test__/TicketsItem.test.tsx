import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import { TicketsItem } from '../TicketsItem';

describe('TicketsItem component', () => {
  afterEach(cleanup);

  const handleDeleteMock = jest.fn();
  const handleEditMock = jest.fn();
  const mockTicket = {
    id: '1',
    title: 'Sample Ticket',
    customerName: 'John Doe',
    customerDate: '2023-10-10T14:30:00Z',
    date: '2023-10-10T14:30:00Z',
    updated: '2023-10-10T14:30:00Z',
    priority: 'high',
    image: 'sample-image-url.jpg',
  };

  const props = {
    handleEdit: handleDeleteMock,
    handleDelete: handleEditMock,
    ticket: mockTicket,
  };

  it('should be rendered', () => {
    const { asFragment } = render(<TicketsItem {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders ticket information correctly', () => {
    render(<TicketsItem {...props} />);

    const titleElement = screen.getByText(mockTicket.title);
    const customerNameElement = screen.getByText(mockTicket.customerName);
    const dateElement = screen.getByText('Oct 10, 2023');
    const priorityLabelElement = screen.getByText(mockTicket.priority);

    expect(titleElement).toBeInTheDocument();
    expect(customerNameElement).toBeInTheDocument();
    expect(dateElement).toBeInTheDocument();
    expect(priorityLabelElement).toBeInTheDocument();
  });

  it('calls handleDelete and handleEdit when speed dial buttons are clicked', () => {
    render(<TicketsItem {...props} />);

    const deleteButton = screen.getByTestId('DeleteIcon');
    const editButton = screen.getByTestId('EditIcon');

    fireEvent.click(deleteButton);
    fireEvent.click(editButton);

    expect(handleDeleteMock).toHaveBeenCalledTimes(1);
    expect(handleEditMock).toHaveBeenCalledTimes(1);
  });
});
