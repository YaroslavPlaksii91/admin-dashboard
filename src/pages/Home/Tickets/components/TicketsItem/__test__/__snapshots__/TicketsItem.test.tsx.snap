// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsItem component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container css-oz5m64-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-b1jvbx-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-2m3hzv"
      >
        <img
          alt="John Doe"
          src="sample-image-url.jpg"
        />
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-k62y85-MuiTypography-root"
        >
          Sample Ticket
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edc1f9-MuiTypography-root"
        >
          Updated today
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-k62y85-MuiTypography-root"
        >
          John Doe
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edc1f9-MuiTypography-root"
        >
          on 10.10.2023
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-k62y85-MuiTypography-root"
        >
          Oct 10, 2023
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edc1f9-MuiTypography-root"
        >
          5:30 PM
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-n268fb-MuiGrid-root"
    >
      <span
        class="undefined"
      >
        high
      </span>
      <div
        class="MuiSpeedDial-root MuiSpeedDial-directionUp css-5jgzwq-MuiSpeedDial-root"
        role="presentation"
      >
        <button
          aria-controls="SpeedDialmoreoptions-actions"
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="SpeedDial more options"
          class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiSpeedDial-fab css-7dv1rb-MuiButtonBase-root-MuiFab-root-MuiSpeedDial-fab"
          style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          tabindex="0"
          type="button"
        >
          <svg
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
          >
            <use
              href="/src/assets/icons/sprite.svg#icon-more"
            />
            <title>
              More options
            </title>
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          aria-orientation="vertical"
          class="MuiSpeedDial-actions MuiSpeedDial-actionsClosed css-1s28u9o-MuiSpeedDial-actions"
          id="SpeedDialmoreoptions-actions"
          role="menu"
        >
          <button
            aria-label="Delete"
            class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiSpeedDialAction-fab MuiSpeedDialAction-fabClosed css-tmh29y-MuiButtonBase-root-MuiFab-root-MuiSpeedDialAction-fab"
            data-mui-internal-clone-element="true"
            role="menuitem"
            style="transition-delay: 60ms;"
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Edit"
            class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiSpeedDialAction-fab MuiSpeedDialAction-fabClosed css-tmh29y-MuiButtonBase-root-MuiFab-root-MuiSpeedDialAction-fab"
            data-mui-internal-clone-element="true"
            role="menuitem"
            style="transition-delay: 30ms;"
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
